import { Injectable, UnauthorizedException } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';

@Injectable()
export class AuthService {
  constructor(private readonly httpService: HttpService) {}

  async login(loginDto: { username: string; password: string }) {
    const url = 'http://localhost:3000/users'; // traemos TODOS los usuarios
    try {
      const response = await firstValueFrom(this.httpService.get(url));
      const users = response.data;

      // Validar que sea un array
      if (!Array.isArray(users)) {
        throw new UnauthorizedException('Respuesta inv치lida del servidor');
      }

      // Buscar el usuario por username
      const found = users.find(
        (u: any) => u.user?.username === loginDto.username,
      );

      if (!found) {
        throw new UnauthorizedException('Usuario o contrase침a incorrecta');
      }

      // Validar contrase침a
      if (found.user.password !== loginDto.password) {
        throw new UnauthorizedException('Usuario o contrase침a incorrecta');
      }

      // Retornamos el objeto user (no todo el documento con _id)
      return { message: 'Login correctamente', user: found.user };
    } catch (error) {
      console.error('Error consultando /users:', error);
      throw new UnauthorizedException('No se pudo validar el usuario');
    }
  }
}
